### PR TITLE
Claim command

### PR DESCRIPTION
**Pull request by Gerrit0**
This adds the `!claim` command, which can be used by mods + members in the TRUSTED_ROLE. It's probably fine to use the helper role for trusted for now. It didn't seem right to let this command be used by *anyone*, but at the same time, I didn't want to restrict it to mods.

Builds on #78.

### CODE REVIEW

**Comment by ronthecookie** [comment]
CodeQL broke but it's not your fault, @robertt any idea why it broke?

**Comment by ronthecookie** [comment]
Looks like it was [reverted](https://github.com/typescript-community/community-bot/commit/f91948e82e25ee911fd647d4e22f9ab5f84414fd#diff-02d9c370a663741451423342d5869b21) but a rebase is required.

**Comment by ronthecookie** [comment]
Ended up doing it myself because I figured out how to avoid the force push

File: src/modules/helpchan.ts [hunk 1]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+		const helpUser = new HelpUser();
+		helpUser.userId = member.user.id;
+		helpUser.channelId = channel.id;
+		await helpUser.save();
+	}
+
+	@command()
+	async cooldown(msg: Message, @optional user?: GuildMember) {
+		console.log('Cooldown', msg.content);
+		if (!msg.guild) return;
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:260]: Use [CommonInhibitors](https://github.com/cookiecord/cookiecord/blob/master/src/command/inhibitors.ts#L18) and the `!` syntax?

File: src/modules/helpchan.ts [hunk 2]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+
+		const guildTarget = await msg.guild.members.fetch(user ?? msg.author);
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:262]: Nice usage of `??` - I don't think all the review comments should be negative.

File: src/modules/helpchan.ts [hunk 3]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+
+		if (!guildTarget.roles.cache.has(askCooldownRoleId)) {
+			await msg.channel.send(
+				`${guildTarget.displayName} doesn't have a cooldown.`,
+			);
+			return;
+		}
+
+		const helpUser = await HelpUser.findOne({
+			where: { userId: guildTarget.id },
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:274]: You don't [need `where`](https://i.ronthecookie.me/3Bk3B6W.png).

File: src/modules/helpchan.ts [hunk 4]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+		if (
+			!msg.member.roles.cache.has(trustedRoleId) &&
+			!msg.member.hasPermission('MANAGE_MESSAGES')
+		) {
+			return msg.channel.send(
+				"You don't have permission to use that command.",
+			);
+		}
+
+		const helpUser = await HelpUser.findOne({ where: { userId: user.id } });
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:305]: You don't [need `where`](https://i.ronthecookie.me/3Bk3B6W.png).

File: src/modules/helpchan.ts [hunk 5]
```diff
@@ -134,21 +138,27 @@ export class HelpChanModule extends Module {
...
-		if (msg.channel.parentID !== categories.ongoing)
+
+		if (msg.channel.parentID !== categories.ongoing) {
 			return await msg.channel.send(
 				':warning: you can only run this in ongoing help channels.',
 			);
+		}
+
+		const owner = await HelpUser.findOne({
+			where: { channelId: msg.channel.id },
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:149]: You don't [need `where`](https://i.ronthecookie.me/3Bk3B6W.png).

File: src/modules/helpchan.ts [hunk 6]
```diff
@@ -193,23 +203,20 @@ export class HelpChanModule extends Module {
...
-			await pinned?.member?.roles.remove(askCooldownRoleId);
-		} else {
-			const helpUser = await HelpUser.findOne({
-				where: { channelId: channel.id },
+
+		const pinned = await channel.messages.fetchPinned();
+		await Promise.all(pinned.map(msg => msg.unpin()));
+
+		const helpUser = await HelpUser.findOne({
+			where: { channelId: channel.id },
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:213]: You don't [need `where`](https://i.ronthecookie.me/3Bk3B6W.png).

**Review by ronthecookie — CHANGES_REQUESTED** [review]

File: src/modules/helpchan.ts [hunk 7]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+		}
+
+		const helpUser = await HelpUser.findOne({ where: { userId: user.id } });
+		if (helpUser) {
+			return msg.channel.send(
+				`${user.displayName} already has an open help channel: <#${helpUser.channelId}>`,
+			);
+		}
+
+		const channelMessages = await msg.channel.messages.fetch({ limit: 50 });
```
[the remark this answers is not in the document]
Comment (Gerrit0) [reply src/modules/helpchan.ts:312]: This was intentional. If any other users reply in between, this command shouldn't miss earlier messages.

[the remark this answers is not in the document]
Comment (ronthecookie) [reply src/modules/helpchan.ts:312]: Ah okay

File: src/modules/helpchan.ts [hunk 8]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
 		}
 	}
 
+	private async addCooldown(member: GuildMember, channel: TextChannel) {
```
[the remark this answers is not in the document]
Comment (Gerrit0) [reply src/modules/helpchan.ts:249]: Any suggestions? I see creating a new `HelpUser` as part of adding a cooldown... not coming up with anything else that isn't just bloating the name.

File: src/modules/helpchan.ts [hunk 9]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+		await member.roles.add(askCooldownRoleId);
+		const helpUser = new HelpUser();
+		helpUser.userId = member.user.id;
+		helpUser.channelId = channel.id;
+		await helpUser.save();
+	}
+
+	@command()
+	async cooldown(msg: Message, @optional user?: GuildMember) {
+		console.log('Cooldown', msg.content);
```
[the remark this answers is not in the document]
Comment (Gerrit0) [reply src/modules/helpchan.ts:259]: Oops ._.

File: src/modules/helpchan.ts [hunk 10]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+			channel =>
+				channel.type === 'text' &&
+				channel.parentID == categories.ask &&
+				channel.name.startsWith(this.CHANNEL_PREFIX) &&
+				!this.busyChannels.has(channel.id),
+		) as TextChannel | undefined;
+
+		if (!claimedChannel) {
+			return msg.channel.send(
+				'Failed to claim a help channel, no available channel.',
```
[the remark this answers is not in the document]
Comment (Gerrit0) [reply src/modules/helpchan.ts]: I added a `:warning:` at the front, which seems to be the main difference

File: src/modules/helpchan.ts [hunk 11]
```diff
@@ -239,6 +246,113 @@ export class HelpChanModule extends Module {
...
+		const toPin = await claimedChannel.send(
+			new MessageEmbed()
+				.setAuthor(user.displayName, user.user.displayAvatarURL())
+				.setDescription(msgContent),
+		);
+		await toPin.pin();
+		await this.addCooldown(user, claimedChannel);
+		await this.moveChannel(claimedChannel, categories.ongoing);
+		await claimedChannel.send(
+			`<@${user.id}> this channel has been claimed for your question. Please review <#${askHelpChannelId}> for how to get help.`,
```
[the remark this answers is not in the document]
Comment (Gerrit0) [reply src/modules/helpchan.ts]: Neat! Done.

File: src/modules/helpchan.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ronthecookie) [reply src/modules/helpchan.ts:249]: I guess it's fine if we don't have any better name ideas.

File: src/modules/helpchan.ts [hunk 12]
```diff
@@ -286,32 +285,20 @@ export class HelpChanModule extends Module {
...
 			return msg.channel.send(
-				`${user.displayName} already has an open help channel: <#${helpUser.channelId}>`,
+				`${member.displayName} already has an open help channel: <#${helpUser.channelId}>`,
 			);
 		}
 
 		const channelMessages = await msg.channel.messages.fetch({ limit: 50 });
 		const questionMessages = channelMessages.filter(
-			msg => msg.author.id === user.id,
+			({ author, id }) => author.id === member.id && id !== msg.id,
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:301]: This is quite confusing to read IMO, maybe no deconstruction and just call it something that doesn't conflict with the existing `msg`?

File: src/util/inhibitors.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import { Inhibitor } from 'cookiecord';
+import { trustedRoleId } from '../env';
+
+export const isTrustedMember: Inhibitor = async (msg, client) => {
+	if (!msg.guild || !msg.member || msg.channel.type !== 'text') {
+		return "You can't use that command here.";
```
Comment (ronthecookie) [inline src/util/inhibitors.ts:6]: Make this message more similar to the rest? (lowercase and `:warning:`)

p.s: also L13

File: src/modules/helpchan.ts [hunk 14]
```diff
@@ -332,23 +319,23 @@ export class HelpChanModule extends Module {
...
 		await toPin.pin();
-		await this.addCooldown(user, claimedChannel);
+		await this.addCooldown(member, claimedChannel);
 		await this.moveChannel(claimedChannel, categories.ongoing);
 		await claimedChannel.send(
-			`<@${user.id}> this channel has been claimed for your question. Please review <#${askHelpChannelId}> for how to get help.`,
+			`${member.user} this channel has been claimed for your question. Please review <#${askHelpChannelId}> for how to get help.`,
 		);
-		await this.ensureAskChannels(msg.guild);
+		await this.ensureAskChannels(msg.guild!);
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:338]: I think you only need `!` on the first reference.

**Review by ronthecookie — CHANGES_REQUESTED** [review]

File: src/modules/helpchan.ts (hunk 14 above)
Comment (Gerrit0) [reply src/modules/helpchan.ts:338]: It's needed everywhere. Assertions don't narrow.

File: src/modules/helpchan.ts [hunk 15]
```diff
@@ -287,6 +285,61 @@ export class HelpChanModule extends Module {
...
+			.join('\n')
+			.slice(0, 2000);
+
+		const claimedChannel = msg.guild!.channels.cache.find(
+			channel =>
+				channel.type === 'text' &&
+				channel.parentID == categories.ask &&
+				channel.name.startsWith(this.CHANNEL_PREFIX) &&
+				!this.busyChannels.has(channel.id),
+		) as TextChannel | undefined;
```
Comment (ronthecookie) [inline src/modules/helpchan.ts:318]: `as` is for asserting the type, use `claimedChannel: TextChan...` instead.

Comment (ronthecookie) [inline src/modules/helpchan.ts:318]: `as` is for asserting, use `claimedChannel: TextChan...` instead.

Comment (Gerrit0) [reply src/modules/helpchan.ts:320]: The assertion is required. `find` doesn't narrow the result to `TextChannel`, so TS thinks it returns `GuildChannel | undefined`

Comment (ronthecookie) [reply src/modules/helpchan.ts:320]: Oh I see now, discord.js has stuff to narrow with`#GuildChannel.type` but that doesn't work with `find`.

**Review by ronthecookie — APPROVED** [review]

**Pull request merged**